### PR TITLE
Hw2

### PR DESCRIPTION
**Pull request by jonathandlee12**

### CODE REVIEW

File: HW2/P2/parallel_vector.pyx [hunk 1]
```diff
@@ -71,20 +79,41 @@ cpdef move_data_fine_grained(np.int32_t[:] counts,
...
+   # &locks is an array of 1000 omp locks, with each lock corresponding to each position in the count array
+   # note: counts begins as a numpy array with index = value
+   # note: src and dest represent arrays of random indices of the counts array
+   
    with nogil:
+       # execute the following script 100 times (repeat = 100)
        for r in range(repeat):
-           for idx in range(src.shape[0]):
+           # loop through each index in src (and dest; total = 1,000,000) - use four threads here
+           for idx in prange(src.shape[0], num_threads=4):
```
Comment (patrickday9) [inline HW2/P2/parallel_vector.pyx:93]: Interesting, you used the prange on the inner loop and I used it on the outer loop. I didn't consider using it on the inner loop as I thought you would want to run parallelly across the repeats. Maybe it makes more sense to implement this way, I wonder which implementation has better performance.

File: HW2/P2/parallel_vector.pyx [hunk 2]
```diff
@@ -71,20 +79,41 @@ cpdef move_data_fine_grained(np.int32_t[:] counts,
...
+                   ind1 = dest[idx]
+                   ind2 = src[idx]
+               else:
+                   ind1 = src[idx]
+                   ind2 = dest[idx]
+
+               # acquire locks accordingly
+               acquire(&locks[ind1])
+	       # check for double-locking
+               if ind1 != ind2:
```
Comment (patrickday9) [inline HW2/P2/parallel_vector.pyx:105]: Cool way to check for checking for double locking once again, I did something slightly different. It's always fun to see different ways of implementing the same thing!

File: HW2/P2/parallel_vector.pyx [hunk 3]
```diff
@@ -95,21 +124,37 @@ cpdef move_data_medium_grained(np.int32_t[:] counts,
...
+               # run the code (need to lock before the conditional to avoid potential negative counts)
                if counts[src[idx]] > 0:
                    counts[dest[idx]] += 1
                    counts[src[idx]] -= 1
 
+               # release the locks
+               release(&locks[ind1])
+               if ind1 != ind2:
+                   release(&locks[ind2])
+
```
Comment (TuanTrain) [inline HW2/P2/parallel_vector.pyx:159]: great!

File: HW2/P4/P4.txt [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+
+To control for cooperation between threads, I chose to use a matrix of events with dimensions of thread number x iterations, where all the events are initialized. As the iterations continue for a particular row n, the program checks to see if the necessary (i-1)th events are available (n-1, n, n+1 threads). This check is skipped for the first iteration. At the end of the script, the event is set to have occurred, and the block is removed, enabling other adjacent threads to run. In particular, this was designed to run asynchronously.
+
+3.48069214821 seconds for 10 filter passes with 1 threads.
+1.89054203033 seconds for 10 filter passes with 2 threads.
+1.75331401825 seconds for 10 filter passes with 3 threads.
+1.7370429039 seconds for 10 filter passes with 4 threads.
+
+As seen in figure P4.png, the speedup from adding a second thread is close to 2; however, successive threads 3 and 4 give very little additional speedup, despite adding multiple threads. I think my computer only has 4 threads (2 cores), but I tried adding extra "threads" to see what happens. There is an extra dip in performance upon the addition of the fifth thread, which levels out to the 4th thread level.
```
Comment (TuanTrain) [inline HW2/P4/P4.txt:10]: It is interesting to see the time for threads beyond the hardware size of your computer is pretty stable

File: HW2/P5/P5.txt [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+Average: 519.855134945 simulation frames per second
+Speedup	 (vs. no lock):	0.71	  
+3 Threads:
+Average: 468.877632942 simulation frames per second
+Speedup	 (vs. no lock):	0.70	  
+4 Threads:
+Average: 556.210696376 simulation frames per second
+Speedup	 (vs. no lock): 0.81
+
+Locking causes a slowdown in performance regardless of the thread count. However, this slowdown still yields a performance much higher than before any of the other manipulations to this code (multithreading, spatial decomposition, sorting). Locking protects our data from multiple collisions being processed improperly.
```
Comment (TuanTrain) [inline HW2/P5/P5.txt:71]: Your results look consistent and thorough!

File: HW2/P3/mandelbrot.pyx [hunk 6]
```diff
@@ -5,9 +5,15 @@ import numpy
 cimport AVX
 from cython.parallel import prange
 
+# helper function to map AVX values in counts to the appropriate positions in out_counts
```
Comment (patrickday9) [inline HW2/P3/mandelbrot.pyx:8]: This is a very clever and clean way of orienting everything in the out_counts, I had a lot of trouble doing this and my solution is not as clean.

File: HW2/P3/mandelbrot.pyx [hunk 7]
```diff
@@ -16,62 +22,45 @@ cpdef mandelbrot(np.complex64_t [:, :] in_coords,
...
+                z_imag = AVX.float_to_float8(0.0)
+
+                # initialize AVX iterate counts to 0.0 and curr to 0.0
+                iters = AVX.float_to_float8(0.0)
+                		
+                # iterate until loop is broken or max_iterations is reached
                 for iter in range(max_iterations):
-                    if magnitude_squared(z) > 4:
+                    # create AVX register containing which elements have |z|^2 < 4
+                    curr = AVX.less_than(AVX.add(AVX.mul(z_real, z_real),AVX.mul(z_imag, z_imag)), AVX.float_to_float8(4.0))
```
Comment (patrickday9) [inline HW2/P3/mandelbrot.pyx:50]: I didn't know you could call AVX functions without saving to a local variable, this looks really clean.

File: HW2/P4/driver.py [hunk 8]
```diff
@@ -17,15 +17,67 @@
...
+        # skip event-checking if this is the first iteration
+        if i != 0:
+            # iterate only among adjacent threads (e.g., other threads are not touched)
+            for j in adj_threads:
+                # for these threads, block until the filtering/switching has occurred
+                events[j, i-1].wait()
+
+        # call median_3x3 Cython code with threadidx-multiple rows and offset num_threads
+        filtering.median_3x3(tmpA, tmpB, threadidx, num_threads)
+        # switch only the pertinent values of tmpA and tmpB
```
Comment (patrickday9) [inline HW2/P4/driver.py:74]: Thank you for including thorough comments, it made this section very easy to understand.

File: HW2/P5/driver.py [hunk 9]
```diff
@@ -75,8 +113,28 @@ def randcolor():
...
+            i += 1
+            if i == iterations:
+                print("Average: {} simulation frames per second".format(total_frame_count / total_total_time))
+                break
+            
+            # sort objects by location using a Morton ordering procedure            
+            grid_positions = (positions / grid_spacing).astype("int")
+            # map grid_positions + offset to the morton ordering function, then obtain the sorted indices
+            grid_indices = np.argsort(map(lambda x, y: cmp_zorder(x, y), grid_positions, np.array(list(grid_positions[1:]) + list(grid_positions[:1]))))
+
```
Comment (patrickday9) [inline HW2/P5/driver.py:129]: Nice one liner for computing the grid indices, this part of the problem took me a long time to figure out. Very clean and easy to follow.

File: HW2/P5/physics.pyx [hunk 10]
```diff
@@ -69,18 +78,37 @@ cdef void sub_update(FLOAT[:, ::1] XY,
...
 
+    grid_size = <unsigned int> (1.0 / grid_spacing + 1)
+    gx = <unsigned int> (XY[i,0] / grid_spacing)
+    gy = <unsigned int> (XY[i,1] / grid_spacing)
+
+    # iterate through a 5 x 5 square on the grid
+    for j in range(gx - 2, gx + 3):
+        for k in range(gy - 2, gy + 3):
+            # ensure j,k are in the grid, and only collide with larger indices (avoid double collisions)
+            if ((0 <= j < grid_size) and (0 <= k < grid_size) and (i < Grid[j,k] < count)):
```
Comment (patrickday9) [inline HW2/P5/physics.pyx:90]: This is a very unique and clever way of checking if j and k are in bounds. I like this implementation of checking if both j and k are greater than 0 and less than the grid size. Very nicely done!

File: HW2/P2/parallel_vector.pyx [hunk 11]
```diff
@@ -71,20 +79,41 @@ cpdef move_data_fine_grained(np.int32_t[:] counts,
...
+               # need to order locks to avoid deadlocking
+               if src[idx] > dest[idx]:
+                   ind1 = dest[idx]
+                   ind2 = src[idx]
+               else:
+                   ind1 = src[idx]
+                   ind2 = dest[idx]
+
+               # acquire locks accordingly
+               acquire(&locks[ind1])
```
Comment (oyuchoi) [inline HW2/P2/parallel_vector.pyx]: by the time you acquire this lock, the true value of dest[idx] and src[idx] may have changed! so its generally a good practice to recheck the condition after acquiring the lock

**Pull request closed without merging**